### ISSUE TITLE
Getting a union from an `and` expression

### ISSUE THREAD

**Issue by Dreamsorcerer** [python/mypy#16565]
**Bug Report**

I'd hope there's already an issue for this, but couldn't find one.
When using `and`, mypy infers a union from both sides when the truthy evaluation makes it impossible.

Actual code trigerring the error: https://github.com/aio-libs/aiohttp/pull/7879/files#diff-720910fd022b9ba8b7ddf0cb446ce5e1afe443f4287ad26a1eb4dc4b5b5f4508R944

**To Reproduce**

```python
class Protocol:
    upgraded: bool

class Connection:
    protocol: Protocol | None

conn: Connection | None
protocol = conn and conn.protocol
assert protocol is not None
protocol.upgraded  # error: Item "Connection" of "Connection | Protocol" has no attribute "upgraded"  [union-attr]
```

**Expected Behavior**

protocol can't be Connection, so this error shouldn't happen.

**Your Environment**

- Mypy version used: 1.7.1

**Comment by erictraut** [comment]
I think mypy is doing the right thing here. The `and` operator terminates early if the LHS evaluates to falsy.

```python
# If `a` is `None`, the result of `a and 8` will be `None`
print(None and 8) # Prints "None"

# If `a` is `Literal[0]`, then `a and 8` will evaluate to `Literal[0]`
print(0 and 8) # Prints "0"

If `a` is any other `int` value, then `a and 8` will evaluate to `Literal[8]`
print(1 and 8) # Prints "8"
```

For comparison, pyright evaluates the type of `a and 8` as `Literal[8, 0] | None`, but mypy's evaluation of `int | None` is also correct.

**Comment by Dreamsorcerer** [comment]
Sorry, I oversimplified it while coming up with the minimal reproducer, it was probably meant to be the inverse of that.

I've just updated the original issue with a slightly more complex example, which better matches the code in the linked PR.

**Comment by Dreamsorcerer** [comment]
To sanity check I've got it right, this example does pass correctly in pyright.

**Comment by erictraut** [comment]
Yeah, I agree that the modified example demonstrates a bug (or at least a missing feature) in mypy.

@AlexWaygood, if you agree, could you reopen the issue?

**Comment by ikonst** [comment]
Couldn't it be a Connection whose `__bool__` returns `False`?

**Comment by erictraut** [comment]
> Couldn't it be a `Connection` whose __bool__ returns `False`?

That's theoretically possible if the class is not `@final`, but the behavior is the same even if it is marked `@final`. That leads me to believe this is a bug, not an intentional design decision. FWIW, the updated sample type checks without errors [in pyright](https://pyright-play.net/?strict=true&code=MYGwhgzhAEAKBOB7ALo4iQC4BQ0-QFcAHAc3jABMBTCzaAI0Q221EhgGFEA7bq4ZAEseOfNCJJU6LHEloM0AD7QAcjyotqAM2haC3YAAp0vOl178hPJavUBKUfgkp5IaAF5oJ7tDDcKXjzcAHTOUsxi7FTwyOJy0tCCMNwotny4%2BPBUAG5UYCAA%2BsgAnkRUhmGuwcRklDR22EA).

**Comment by ikonst** [comment]
It appears that mypy isn't smart enough to deduce "if @final without __bool__ or __len__, then always true", but otherwise adding `__bool__` works as expected. Should we make mypy smarter?

As for a workaround, either seems to work

1. adding a `__bool__`
   ```diff
    class Connection:
      protocol: Protocol | None
   +  def __bool__(self) -> Literal[True]: return True
   ```
   No need to make it `@final` since derived class must be covariant on return type.

2. letting mypy know you never want a falsey Connection:
   ```diff
    def func(conn: Connection | None):
   -  protocol = conn and conn.protocol
   +  protocol = (conn or None) and conn.protocol
      assert protocol is not None
      reveal_type(protocol.upgraded)
   ```

**Comment by Dreamsorcerer** [comment]
> 2\. letting mypy know you never want a falsey Connection:

Would there be an easy way to make this suggestion in the error output? Our example is actually not Final, we just didn't think of that edge case.

**Comment by ikonst** [comment]
Assuming you don't want to change `Connection` i.e. you want to address it at the call site, then "suggestion (2)" would be relevant.

`protocol = conn and conn.protocol` itself isn't an error. Does it on its own warrant a warning? Probably not, since it could actually be your intention, e.g.
```python
def make_request() -> requests.Response | dict:
  resp = requests.get('http://example.com')
  return resp and resp.json()
```
Maybe `x and y` should only warn if `x` is `DontKnowIfTruthy | None` and you end up assigning it to a variable rather than just use it in boolean context... I don't know how easy it is to implement, and how noisy it'll be.

Keeping context and suggesting when the `union-attr` actually happens... sounds pretty complicated, and I can't even think how you pharse it.

**Comment by Dreamsorcerer** [comment]
We decided to go with the first option, as we may have made similar assumptions elsewhere in the code and it keeps the logic simple. Both solutions would be useful to suggest in this context.

> Keeping context and suggesting when the `union-attr` actually happens... sounds pretty complicated, and I can't even think how you pharse it.

Yeah, I figured that might be difficult as you'd need to know the cause was from the code further up.

**Comment by ikonst** [comment]
#16566 was merged so starting from mypy 1.8(?) adding `@final` should be enough.
